Adding simple Drizzle Node.js example (#89)
MaterializeInc/demos: integrations/nodejs/drizzle/README.md

In this demo, we'll explore how to integrate Node.js with Drizzle ORM to query data from Materialize, a streaming database.

### Prerequisites:

-   Node.js installed on your machine
-   Materialize instance
-   Materialize [getting started guide](https://materialize.com/docs/get-started/)

### Step 1: Setting up the Project

Begin by setting up a new Node.js project:

```bash
mkdir drizzle-materialize-demo
cd drizzle-materialize-demo
npm init -y
```

### Step 2: Install Dependencies

Install the required packages:

```bash
npm install drizzle-orm pg dotenv
```

### Step 3: Configuration

Duplicate the provided `.env.example` into a new `.env` file:

```bash
cp .env.example .env
```

Update the `.env` file with your Materialize credentials.

### Step 4: Establishing a Connection

Let's understand the given code in `app.js`:

```js
import dotenv from "dotenv";
import pkg from "pg";
const { Client } = pkg;

dotenv.config();

// Create a connection client
const client = new Client({
  host: process.env.MZ_HOST,
  port: Number(process.env.MZ_PORT),
  user: process.env.MZ_USER,
  password: process.env.MZ_PASSWORD,
  database: process.env.MZ_DATABASE,

>   ssl: {

Reviewer: Why is it not just `sll: true`?

Author: Good point! Just updated this

    rejectUnauthorized: false,
  },
});

await client.connect();
```

This code does the following:

1.  Imports necessary modules.
2.  Reads environment variables from the `.env` file.
3.  Creates a PostgreSQL client and connects to the Materialize database.

### Step 5: Defining the Data Structure with Drizzle

Drizzle provides an ORM (Object-Relational Mapping) interface. The given code defines a view that maps to the `winning_bids` view in the database:

```js
import { pgView, serial, text, integer, real, timestamp } from "drizzle-orm/pg-core";
import { drizzle } from "drizzle-orm/node-postgres";

const db = drizzle(client);

> export const winningBids = pgView("winning_bids", {

Reviewer: Does `pgView` works with Materialized Views?

Reviewer: Ohh. I saw that there is also `pgMaterializedView`

Reviewer: When I try to run this code, with a materialized view, the app throws me:

```
file:///Users/joaquincolacci/Code/Materialize/demos/integrations/nodejs/drizzle/node_modules/drizzle-orm/pg-core/index.mjs:48
    if (config.mode === 'number') {
               ^

TypeError: Cannot read properties of undefined (reading 'mode')
```

There are high chances that I might be doing something wrong.

Author: This is an interesting idea! My initial idea was to ask people to run through the getting started demo listed in the prerequisites.

But we could also add this in the script and automate it, which one do you think is better?

Reviewer: > This is an interesting idea! My initial idea was to ask people to run through the getting started demo listed in the prerequisites.

If this is the initial thought, then I'd say to add more details and link/ref (besides the prereqs.). The get started docs have a clean up step at the end, so chances that the table/view being is not available is high!

Author: Added a note to mention that users should follow the getting started demo but skip the cleanup steps. In case that anyone tries to run the example script the error should also be quite self-explanatory eg:

```js
error: unknown catalog item 'winning_bids'
```

    id: serial("id"),
    buyer: text("buyer"),
    auction_id: integer("auction_id"),
    amount: real("amount"),
    bid_time: timestamp("bid_time"),
    item: text("item"),
    seller: text("seller")
}).existing();
```

Here, `pgView` is used to map the database view to a JavaScript object. Data types like `serial`, `text`, `integer`, etc., are used to define the columns and their types.

> ### Step 6: Fetching Data

Reviewer: Since the project does not have a `start` declared in the package. We should have a comment in one step about how to run the code. E.g. `node app.js`. Maybe after the install.

Author: Good catch! I added an extra step for this

## Node.js + Drizzle + Materialize demo

> In this demo, we'll explore how to integrate Node.js with Drizzle ORM to query data from Materialize, a streaming database.

Reviewer 2: ```suggestion
In this demo, we'll explore how to integrate Node.js with Drizzle ORM to query data from Materialize, an operational data warehouse.
```
